### PR TITLE
Migrate TF1 to TF2

### PR DESCRIPTION
**Pull request by RoyStegeman**
Here we migrate `evolutionary_keras` from TF1 to TF2

### CODE REVIEW

**Comment by scarlehoff** [comment]
Just checking, I guess tests and pylint is passing ok in your computer. Which versions of Tensorflow and python are you using?

**Comment by RoyStegeman** [comment]
Yes they're passing. 

Python 3.7.6
Tensorflow 2.1.0

**Comment by scarlehoff** [comment]
> Tensorflow 2.1.0

It doesn't for the RC of 2.2. Let's see whether there is an easy fix.

**Comment by scarlehoff** [comment]
Also, just noticed that you left keras imports in `test_utilities.py`

**Comment by RoyStegeman** [comment]
> It doesn't for the RC of 2.2. Let's see whether there is an easy fix.

I can try to find a fix. But doesn't it make more sense to wait for the final release?

> Also, just noticed that you left keras imports in test_utilities.py

Yes, there was a problem with the tf.keras layer. Even after calling it with `layer(ii)`, when trying to get the number of notes, it gives an error that the layer has not yet been called. Perhaps I'll have to initiate a complete model there as well. I had kinda forgotten about the tests when I asked your review :disappointed:

**Comment by scarlehoff** [comment]
> I can try to find a fix. But doesn't it make more sense to wait for the final release?

Yes, but it makes sense to think on future proofing on 2.2 rather than 2.1 because it brings in Keras 2.3. Keras 2.3 is the last version of Keras developed outside of TF so it is to be expected that it will be more stable from there onwards https://keras.io/

**Comment by scarlehoff** [comment]
Tricky, the error with 2.2 was a problem that I don't think should be. Basically what's happening is that running `evaluate` is resetting `history` to a point before the training started, so the list of epochs does not exist. It might be a bug within 2.2 but in any case, 2.2 allows to stop using such as `parse_eval` as they have unified how evaluate and fit behave.

So I'd say the way to go around this is:
1. Fix the problem with `test_utilities`
2. Fix outstanding problems with pylint and so on
--> 2.1 compatible
3. Wait for 2.2 to be finally out (polish things like `parse_eval` in the meantime)
4. Make it sure it works with 2.2, create a new release, create a new conda package
5. Go to the NNPDF repository and merge it to master <---- we'll be careful here and not substitute `keras_backend` but rather we'll create an `evolutionary_backend`, but let's not worry about this until  1 to 4 are done, we have no hurry.

I'll push a workaround so it works with the release candidate of 2.2

**Comment by RoyStegeman** [comment]
> 2.2 allows to stop using such as parse_eval as they have unified how evaluate and fit behave.

I expected that this meant they had evaluate output a dict as well, but it's still a list, hence we still need `parse_eval`. Or what do you mean with 'they unified the output'? I can't find it among the list of changes with the 2.2 release either.

**Comment by scarlehoff** [comment]
> There aren't. Not locally at least.

It's a problem known by both the pylint and tf developers and in theory it is fixed in tf>2.2

> I expected that this meant they had evaluate output a dict as well, but it's still a list,

You have to ask for it by adding the `return_dict` keyword argument https://www.tensorflow.org/versions/r2.2/api_docs/python/tf/keras/Model#evaluate

File: src/evolutionary_keras/optimizers.py [hunk 1]
```diff
@@ -83,7 +80,18 @@ def __init__(
         self.n_nodes = 0
         self.n_generations = 1
 
-        super(NGA, self).__init__(*args, **kwargs)
+        super(NGA, self).__init__(name, **kwargs)
```
Comment (scarlehoff) [inline src/evolutionary_keras/optimizers.py:83]: Is there a reason why *args are not passed here? I think it should go in as well...

Comment (RoyStegeman) [reply src/evolutionary_keras/optimizers.py:94]: hmm, not really sure. I think the tf optimizers don't take *args inputs anyway, so we might as well get rid of it. Not sure why it is still in the init of NGA itself or why we pass the name though. 

I don't see any harm in passing args though, so for now we might as well.

Comment (scarlehoff) [reply src/evolutionary_keras/optimizers.py:94]: Well, the base Optimizer does. Actually we should pass the lock as well... I'll do that  https://github.com/tensorflow/tensorflow/blob/e3e0ba57815094e58c6df16ef1b68281418bf4a6/tensorflow/python/training/optimizer.py#L313

Comment (scarlehoff) [reply src/evolutionary_keras/optimizers.py:94]: Oh, I was looking a TF 1. TF 2 takes only name, then you are right, *args should be removed.

Comment (RoyStegeman) [reply src/evolutionary_keras/optimizers.py:94]: Oh I see, you're right. Sorry about that.

**Pull request merged**